File: packages/ts/tests/binding/extension-package.test.ts
// Binding tier: resolve a SQLite extension by bare package name (#299).
//
// Lays a real compiled loadable inside a real installed-package directory under
// `node_modules`, points `DirSQL` at it by **bare package name**, and asserts
// the SDK resolves the actual on-disk file via `require.resolve`, loads it, and
// the function it registers is callable. Real layout, no mocks -- the shape
// mirrors the Python sibling (#298) and the Rust end-to-end test: resolve ->
// load -> callable.
//
// The loadable is the repo's `tests/fixtures/testext` cdylib (registers
// `dirsql_testext_answer() -> 42`), built on the fly with cargo.

import { execFileSync } from "node:child_process";
import { mkdirSync, mkdtempSync, rmSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { join, resolve } from "node:path";
import { DirSQL } from "dirsql";
import { afterEach, beforeEach, describe, expect, it } from "vitest";

const repoRoot = resolve(import.meta.dirname, "..", "..", "..", "..");
const tsNodeModules = resolve(import.meta.dirname, "..", "..", "node_modules");
const fixtureManifest = join(
  repoRoot,
  "packages/rust/tests/fixtures/testext/Cargo.toml",
);

function buildFixtureExtension(targetDir: string): string {
  const stdout = execFileSync(
    process.env.CARGO ?? "cargo",
    [
      "build",
      "--manifest-path",
      fixtureManifest,
      "--target-dir",
      targetDir,
      "--message-format=json",
    ],
    { encoding: "utf8", maxBuffer: 64 * 1024 * 1024 },
  );
  let artifact: string | undefined;
  for (const line of stdout.split("\n")) {
    if (!line) {
      continue;
    }
    let msg: { reason?: string; filenames?: string[] };
    try {
      msg = JSON.parse(line);
    } catch {
      continue;
    }
    if (msg.reason === "compiler-artifact" && msg.filenames) {
      for (const f of msg.filenames) {
        if (f.endsWith(".so") || f.endsWith(".dylib") || f.endsWith(".dll")) {
          artifact = f;
        }
      }
    }
  }
  if (!artifact) {
    throw new Error("no cdylib artifact in cargo build output");
  }
  return artifact;
}

describe("DirSQL extension by package name (#299)", () => {
  // Must stay unique across test files: suites run in parallel workers and
  // share this `node_modules`, so a reused name races on setup/cleanup (#349).
  const pkgName = "dirsql-testext-pkg";
  const pkgDir = join(tsNodeModules, pkgName);
  let tmp: string;

  beforeEach(() => {
    tmp = mkdtempSync(join(tmpdir(), "dirsql-ext-pkg-"));
    const so = buildFixtureExtension(join(tmp, "target"));
    // A real installed-package layout under the SDK's node_modules so
    // `require.resolve` finds it.
    mkdirSync(pkgDir, { recursive: true });
    writeFileSync(
      join(pkgDir, "package.json"),
      JSON.stringify({ name: pkgName, version: "0.0.0" }),
    );
    execFileSync("cp", [so, join(pkgDir, "testext.so")]);
  });

  afterEach(() => {
    rmSync(pkgDir, { recursive: true, force: true });
    rmSync(tmp, { recursive: true, force: true });
  });

  it("resolves, loads, and calls an extension referenced by package name", async () => {
    const db = new DirSQL({
      root: tmp,
      extensions: [{ path: pkgName, entrypoint: "sqlite3_extension_init" }],
    });
    await db.ready;
    expect(await db.query("SELECT dirsql_testext_answer() AS a")).toEqual([
      { a: 42 },
    ]);
  });
});

File: packages/python/tests/binding/extension_package_test.py
"""Binding tier: resolve a SQLite extension by bare package name (#298).

Lays a real compiled loadable inside a real package directory on `sys.path`,
points `DirSQL` at it by **bare package name**, and asserts the SDK resolves
the actual on-disk file, loads it, and the function it registers is callable.
Real layout, no mocks -- the shape mirrors the TypeScript sibling (#299) and
the Rust end-to-end test: resolve -> load -> callable.

The loadable is the repo's `tests/fixtures/testext` cdylib (registers
`dirsql_testext_answer() -> 42`), built on the fly with cargo.
"""

import importlib
import json
import os
import shutil
import subprocess
import sys

import pytest

from dirsql import DirSQL

_REPO_ROOT = os.path.abspath(
    os.path.join(os.path.dirname(__file__), "..", "..", "..", "..")
)
_FIXTURE_MANIFEST = os.path.join(
    _REPO_ROOT, "packages", "rust", "tests", "fixtures", "testext", "Cargo.toml"
)


def _build_fixture_extension(target_dir):
    """Build the testext cdylib and return the path to its loadable file."""
    proc = subprocess.run(
        [
            os.environ.get("CARGO", "cargo"),
            "build",
            "--manifest-path",
            _FIXTURE_MANIFEST,
            "--target-dir",
            target_dir,
            "--message-format=json",
        ],
        capture_output=True,
        text=True,
        env={
            k: v
            for k, v in os.environ.items()
            if k not in ("RUSTFLAGS", "RUSTDOCFLAGS")
        },
    )
    assert proc.returncode == 0, f"fixture build failed:\n{proc.stderr}"
    artifact = None
    for line in proc.stdout.splitlines():
        try:
            msg = json.loads(line)
        except ValueError:
            continue
        if msg.get("reason") == "compiler-artifact":
            for f in msg.get("filenames") or []:
                if f.endswith((".so", ".dylib", ".dll")):
                    artifact = f
    assert artifact, "no cdylib artifact in cargo build output"
    return artifact


def describe_extension_by_package_name():
    @pytest.mark.asyncio
    async def it_resolves_loads_and_calls_an_extension_by_package_name(tmp_path):
        if shutil.which(os.environ.get("CARGO", "cargo")) is None:
            pytest.skip("cargo not available to build the extension fixture")

        so = _build_fixture_extension(str(tmp_path / "target"))

        # A real installed-package layout:
        # <env>/dirsql_testext_pkg/{__init__.py, *.so}
        env = tmp_path / "env"
        pkg = env / "dirsql_testext_pkg"
        pkg.mkdir(parents=True)
        (pkg / "__init__.py").write_text("")
        shutil.copy(so, pkg / os.path.basename(so))

        sys.path.insert(0, str(env))
        importlib.invalidate_caches()
        try:
            db = DirSQL(
                str(tmp_path),
                extensions=[
                    {
                        "path": "dirsql_testext_pkg",
                        "entrypoint": "sqlite3_extension_init",
                    }
                ],
            )
            await db.ready()
            assert await db.query("SELECT dirsql_testext_answer() AS a") == [{"a": 42}]
        finally:
            sys.path.remove(str(env))
            sys.modules.pop("dirsql_testext_pkg", None)
            importlib.invalidate_caches()
